### PR TITLE
add depth limiting

### PR DESCRIPTION
**Pull request by rickbijkerk**
Add config option to configure max allowed depth for queries:

example config:
```
max_depth:
  enable: true
  # The maximum number of allowed aliases within a single request.
  max: 3
  # Reject the request when the rule fails. Disable this to allow the request
  reject_on_failure: true
```

For example the query below has a depth of 4.
```
query{
  activeExperiments {
    variants {
      experimentKey
      allocations {
        end
      }
    }
  }
}

```

With the config above it would fail with error:
```

{
  "data": {},
  "errors": [
    {
      "message": "syntax error: Depth limit of 3 exceeded, found 4",
      "locations": [
        {
          "line": 1,
          "column": 1
        }
      ]
    }
  ]
}
```

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -69,6 +69,13 @@ max_aliases:
 block_field_suggestions:
   enabled: true
   mask: [redacted]
+  
+max_depth:
+  enable: true
+  # The maximum number of allowed aliases within a single request.
```
Comment (ldebruijn) [inline docs/configuration.md:75]:
```suggestion
  # The maximum allowed depth within a single request.
```

File: internal/business/max_depth/max_depth.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+func init() {
+	prometheus.MustRegister(resultCounter)
+}
+
+func NewMaxDepthRule(cfg Config) {
+	if cfg.Enabled {
+		validator.AddRule("MaxDepth", func(observers *validator.Events, addError validator.AddErrFunc) {
+
```
Comment (ldebruijn) [inline internal/business/max_depth/max_depth.go:32]: empty line

File: internal/business/max_depth/max_depth.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+				depth = selectionDepth
+			}
+		}
+	}
+	return depth
+
+}
+
+//
+//func countDepth(set ast.SelectionSet) int {
```
Comment (ldebruijn) [inline internal/business/max_depth/max_depth.go:81]: remove this please

**Review by ldebruijn — CHANGES_REQUESTED** [review]
Thanks for your contribution! Great protection to have in our feature set :)

File: internal/business/max_depth/max_depth.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
+package max_depth
+
+import (
+	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/vektah/gqlparser/v2/ast"
+	"github.com/vektah/gqlparser/v2/validator"
+)
+
+var resultCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
```
[the remark this answers is not in the document]
Comment (rickbijkerk) [reply internal/business/max_depth/max_depth.go:10]: i'll leave that one for another time

**Review by ldebruijn — APPROVED** [review]

**Pull request merged**